### ISSUE TITLE
WO-014: single-enterprise DP

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#19]
> **WO-014** - Phase 1 - Owner **MID-strong** - Difficulty **4/5** - Gate **G1**
> Card: [`workorders/WO-014.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-014.md)

## What this is

This card solves the environment exactly, for one enterprise, by dynamic programming. Strip the
system down to a single enterprise with no input-output linkages and fixed capacity, and the
enterprise's problem becomes a two-dimensional Markov decision process over (target, stock) that can
be solved by value iteration on a grid. The result is ground truth: the optimal effort and report
policy, the stationary distribution of the reported fulfilment ratio, and the excess mass
`b_hat_DP` that a correct learner should reproduce.

It exists for two reasons. First, it is the yardstick for gate G2 criterion 1: if PPO cannot recover
this policy at N = 1, the training stack is broken and nothing downstream means anything. Second, it
is the analytical layer that the regime map (WO-015) sweeps, so the Phase-1 parameter values can be
chosen from an exactly-solved map BEFORE any learning run, and therefore cannot be chosen to suit a
training result.

## Depends on

- `WO-013`
- gate G0, transitively: WO-013 does not start until G0 is signed off, so this card inherits that
  requirement (PLAN section 12.2). `spec/spec.py` is v1.0.0 and frozen when this card runs.

**Blocked by:** #18 (WO-013), #43 (gate G0)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- `PLAN.md` - SECTION 5 IN FULL, the recipe this card implements, plus sections 2.6 (yield and
  cost), 2.7.1 (target rule), 2.8 (bonus, audit, penalty), 2.9.1 (`reward_scale`), 2.11 (holding
  loss, `S_max`), 2.12 (geometric continuation), 4.1 rows 1 and 4, 4.4 (measurement window), 4.5
  (the pre-registered estimator settings `b_hat_DP` uses), 6.1 (`DPGreedy`), 11, 12.3 (this card),
  13 (gate G1), 14 (cost). `PLAN.md` is git-ignored and never committed; it must be present at the
  repository root locally, supplied out of band.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rules 1, 3, 4, 8,
  10, 12
- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - v1.0.0, the
  frozen `DPGrid`, `DPSolution`, `RegimeLabel` and the DP function signatures
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py) -
  `EnvConfig` and every incentive field the problem reads
- [`gosplan/params.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/params.py) - the
  registry behind those defaults, including which rows are provisional until G1
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py) - read for the
  seed conventions; the DP's own simulation seeds
- [`gosplan/env/reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reward.py) -
  `bonus`, `reward_scale` - CALL THESE, do not re-derive them
- [`gosplan/env/reporting.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reporting.py)
  - the PLAN section 2.8 audit and penalty definitions
- [`gosplan/agents/dp.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/dp.py) -
  WRITE TARGET. `DP_DISCOUNT`, `DPGrid` and `DPSolution` field declarations are already there; eight
  bodies missing
- [`gosplan/agents/base.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/base.py)
  - the `Agent` protocol
- [`gosplan/agents/heuristic.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/heuristic.py)
  - `DPGreedy`, the consumer of `DPSolution` (WO-010's; read only)
- [`gosplan/metrics/phenomena.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/metrics/phenomena.py)
  - THE `BUNCHING_*` PLAN section 4.5 CONSTANTS ONLY, for `b_hat_DP`; the Phase-1 bodies land at
  WO-016
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - ruff
  line-length 100, target py312
- [`tests/unit/test_dp.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/unit/test_dp.py) -
  read-only (rule 2), from WO-002

Nothing else. Not `gosplan/agents/ppo/`, not `gosplan/experiments/`, not `gosplan/env/step.py`.
Read nothing else. CONTRACT rule 12.

## Write only

- `gosplan/agents/dp.py` - `solve_single_enterprise`, `classify_regime`,
  `gauss_hermite_lognormal_nodes`, `report_action_grid`, `report_grid_edge_fraction`,
  `simulate_stationary_reports`, `dp_excess_mass`. `DP_DISCOUNT` and the `DPGrid` PLAN section 5
  values keep their declared values

`DPGreedy` lives in `gosplan/agents/heuristic.py` and is WO-010's. If it needs a change, that is an
AMBIGUITY REPORT, not an edit.

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

- [ ] `tests/unit/test_dp.py` - value iteration converges
- [ ] `tests/unit/test_dp.py` - with `a*pen -> inf` and `g = 0` the policy reports truthfully
- [ ] `tests/unit/test_dp.py` - with `beta = 0`, `s = 0` effort is 0
- [ ] `tests/unit/test_dp.py` - the regime classifier on synthetic distributions
- [ ] `tests/unit/test_dp.py` - `DPGreedy` reproduces the DP policy inside the env at `N = 1`
- [ ] The full frozen suite stays green

```bash
uv run ruff check gosplan/agents/dp.py
uv run ruff format --check gosplan/agents/dp.py
uv run pytest tests/unit/test_dp.py -q
uv run pytest -q
```

## Implementation notes

All of the following is PLAN section 5, transcribed. Cite it in the code.

**Setting.** One enterprise; NO input-output (`a = 0`, `phi = 1`), so coverage is 1 and every claimed
unit can be delivered; fixed capacity; geometric continuation `psi = cfg.incentive.tenure`.

**State grid.**

```
T in [T_min, grid.target_hi_mult * A * cap]      80 points, LOG-SPACED
    T_min = cfg.tech.target_floor_frac * T_0
    T_0   = cfg.tech.initial_target_frac * A * cap
S in [0, S_max]                                  50 points, LINEAR
    S_max = cfg.tech.inventory_cap_mult * cap
```

**Action grid.**

```
effort e   in {0, 0.05, ..., 1}      21 points, ONE EFFORT PER PERIOD
report rho in [0, 3], step 0.02      extended automatically at the edge
```

One effort per period, not per step: nothing within the period is observed, so uniform effort across
the `M` steps is optimal and its cost is `M * kappa * e**2`. The report grid is EXTENDED
AUTOMATICALLY when the optimum sits at the edge, and every edge hit is LOGGED AS A REGIME SIGNAL
(`report_action_grid`, `report_grid_edge_fraction`, `DPSolution.rho_edge_frac`), never smoothed away
(CONTRACT rule 8).

**Period yield.**

```
y = A * cap * e * eps_bar
eps_bar ~ LogNormal(-sigma_bar**2 / 2, sigma_bar),   sigma_bar = sigma / sqrt(M)
```

This is the aggregation approximation and MUST BE DOCUMENTED IN THE CODE AS AN APPROXIMATION. Take
the expectation with 9-node GAUSS-HERMITE IN LOG SPACE (`grid.gh_nodes`);
`numpy.polynomial.hermite.hermgauss` supplies nodes and weights, so no `scipy` at module scope.

**Transition.**

```
S'  = (1 - h) * S + y
R   = rho * T
audit against S' with probability a and log-noise sigma_aud:
        S_hat = S' * exp(nu),  nu ~ N(0, sigma_aud**2)
S'' = S' - min(S', R)                                  # delivery next period, phi = 1
T'  = max(T_min, (1 + g) * T * (1 + lambda * clip(rho - 1, -c_dn, c_up)))
```

**Bellman operator.**

```
V(T, S) = max_{e, rho} { E_eps[ B(rho) - M * kappa * e**2 - a * E_nu[Pen] ]
                         + psi * gamma * E V(T', S'') }
```

with `gamma = DP_DISCOUNT = 0.99`, the TECHNICAL discount, identical to `PPOConfig.gamma` because G2
criterion 1 compares the two policies directly and a discount mismatch would compare two different
problems. `psi` is the ECONOMIC continuation probability `cfg.incentive.tenure`. They multiply here
and are conflated nowhere else. Value iteration to `grid.value_tol = 1e-6`, or policy iteration -
either is acceptable provided it reports the same `converged` flag and `n_iterations`. Do not invent
a different convergence criterion (CONTRACT rule 3), and do not add a term the operator does not
have: `B(rho)`, the effort cost and the expected penalty, and nothing else (CONTRACT rule 4).

**Use the environment's own functions.** `B(rho)` is `gosplan.env.reward.bonus(rho, cfg)` and the
scale is `gosplan.env.reward.reward_scale(cfg)` - CALL THEM. Do not re-derive the notch, the
overfulfilment slope or the cap: `Lambda_w` with `w = 0` is a strict `>=` Heaviside and
`rho_cap = inf` means no clip, and those branches are already right there. This is also what keeps
the regime map (WO-015) and the environment from ever drifting apart.

**The penalty is the one exception.** `audit_and_penalise` takes a `State`, which the DP does not
have. Evaluate the PLAN section 2.8 penalty on the DP's scalar state, covering BOTH `penalty_form`
branches and BOTH `penalty_arg` branches:

```
penalty_form:  pen * f            or    pen * 1[f > 0]
penalty_arg:   f = max(0, R - S_hat) / T    or    f = |R - S_hat| / T
```

with a comment citing PLAN section 2.8. DO NOT REFACTOR `gosplan/env/reporting.py` - it is not on
your "Write only" list. If the two definitions could disagree, that is an AMBIGUITY REPORT, never a
silent divergence. This is the likeliest ambiguity on the card.

**Outputs - every field of `DPSolution`.**

- `policy_effort` and `policy_rho` on the `(n_target, n_stock)` grid
- `value`
- the stationary distribution of `rho_report` from simulating the policy 200 EPISODES x 200 PERIODS
  (`simulate_stationary_reports`, burn-in per PLAN section 4.4)
- `b_hat_dp` (`dp_excess_mass`), computed with the PRE-REGISTERED PLAN section 4.5 settings: bins of
  width 0.005 on `rho` in [0.6, 1.4], excluded window [0.95, 1.02], polynomial degree 7, excess mass
  on [1.00, 1.02]
- `fictitious_padding` = `mean max(0, R - S) / T`
- `hidden_reserves` = `mean max(0, S - R) / T` after delivery
- `mean_effort`
- `regime`
- `rho_edge_frac`
- `n_iterations`
- `converged`
- `grid`
- `config_hash = cfg.hash()`

**`classify_regime` - thresholds verbatim (PLAN section 5):**

```
bunching                  mass of rho in [1.00, 1.005] > 0.5 AND mass at the grid edge < 0.05
pad_to_cap                mass at the grid edge > 0.5
truthful_underfulfilment  mean rho < 0.9 AND fictitious padding < 0.01
mixed                     otherwise
```

Do not relax a threshold, and do not re-grid after an edge hit to make a label come out prettier
(CONTRACT rule 8).

**Performance.** Vectorise over the action grid (the `(e, rho)` outer product against the state
grid). PLAN section 14 budgets about one CPU-minute per configuration at this resolution, and WO-015
solves 500 of them. Report the observed solve time for one `p1_default_config()` solve against that
budget.

**Non-convergence is a result.** A solution with `converged = False` is returned WITH THE FLAG and is
never silently used as ground truth at gate G2. The same holds for `rho_edge_frac`: an edge hit is a
regime signal, logged, never re-gridded away.

**No interpolation in the returned policy tables.** `DPGreedy` reads them by nearest grid point, and
the fifth must-pass item depends on exact table entries.

## Forbidden

- ANY REINFORCEMENT LEARNING. This is the WO-014 forbidden line of PLAN section 12.3, verbatim: no
  policy gradient, no PPO, no Q-learning, no sampling-based policy search, no learned function
  approximation, no neural network anywhere in this file. The analytical layer is solved BEFORE ANY
  MARL precisely so the Phase-1 values cannot be chosen to suit a training result.
- Re-implementing `bonus`, `reward_scale` or the target rule instead of calling or transcribing the
  frozen definitions; editing `gosplan/env/` or `gosplan/agents/heuristic.py`.
- Widening a grid, loosening `value_tol`, or dropping an edge-hit configuration to obtain a cleaner
  regime label (rule 8).
- Adding a term to the Bellman operator that is not in CONTRACT rule 4: `B(rho)`, the effort cost and
  the expected penalty, and nothing else. No shaping, no curiosity, no potential-based term, no
  running normalisation.
- Interpolating between grid points in the returned policy tables.
- Reporting, plotting or tabulating `hidden_reserves` anywhere in Phase 1. It is PLAN section 4.1
  row 7, HELD OUT until the Phase-2 acceptance run: the field is computed and stored, never
  displayed.
- Importing `scipy`, `pandas`, `matplotlib` or `forensics_core` at module scope.
- Touching `tests/`, `ref/`, `spec/spec.py` or `PLAN.md`. Running `tests/acceptance/` (rule 13).

## Contract rules that bite here

- **Rule 3 (STOP AND REPORT).** No invented interpolation scheme, no invented convergence criterion,
  no extra term in the Bellman operator. Choosing "the reasonable default" is a violation: file an
  AMBIGUITY REPORT (`workorders/AMBIGUITY_TEMPLATE.md`) and end the session.
- **Rule 1 (frozen spec).** `spec/spec.py` is v1.0.0 from WO-013 onwards. Nothing here may change it;
  a needed change is an AMBIGUITY REPORT to the lead.
- **Rule 2 (frozen tests).** `tests/unit/test_dp.py` is read-only. Do not edit, skip, xfail or
  special-case the implementation to pass it.
- **Rule 4 (reward terms).** The Bellman operator carries the reward terms and nothing else.
- **Rule 8 (bounds are results).** An `rho` grid-edge hit is logged as a regime signal, never
  re-gridded away.
- **Rule 10 (manifest).** `config_hash` and the `converged` flag travel with the solution so a result
  is always traceable to the configuration that produced it.
- **Rule 12 (whitelist).** Read only the listed files, write only `gosplan/agents/dp.py`, run the
  completion command verbatim, report in the format above.
- **Rule 13 (tests are not experiments).** Do not run `tests/acceptance/`.

## Definition of done

- [ ] All five `tests/unit/test_dp.py` items green
- [ ] Full frozen suite (`uv run pytest -q`) green
- [ ] `uv run ruff check . && uv run ruff format --check .` clean
- [ ] `uv run python scripts/contract_guard.py --all` clean
- [ ] Observed solve time for one `p1_default_config()` solve recorded against the PLAN section 14
      one-minute budget
- [ ] No file outside "Write only" touched - expect exactly `gosplan/agents/dp.py`
- [ ] CI `ci-ok` green on the PR
- [ ] The four-part report posted as the PR body: (1) tests passed/failed with output -
      `tests/unit/test_dp.py` verbatim, all five items, plus the full frozen suite and the solve
      time; (2) ambiguity reports filed, or "none" - the penalty-definition duplication is the most
      likely one; (3) files changed; (4) anything you were tempted to invent and did not -
      interpolation schemes, alternative convergence criteria, an extra Bellman term, a re-grid after
      an edge hit, or any use of a learned approximation
